compositor_pipeline/src/pipeline/input/mp4.rs

```diff
@@ -0,0 +1,301 @@
+
+    let sample_count = track.sample_count();
+    let timescale = track.timescale();
+    let length_size = avc1.avcc.length_size_minus_one + 1;
+
+    let thread = std::thread::Builder::new()
+        .name(format!("mp4 reader {input_id}"))
+        .spawn(move || {
+            reader_thread(
+                sps_and_pps_payload,
+                reader,
+                sender,
+                stop_thread,
+                length_size,
+                TrackInfo {
+                    sample_count,
+                    timescale,
+                    track_id,
+                },
+            )
+        })
+        .unwrap();
+
+    Ok((thread, receiver))
+}
+
+struct TrackInfo {
+    sample_count: u32,
+    track_id: u32,
+    timescale: u32,
+}
+
+fn reader_thread(
+    mut sps_and_pps: Option<Bytes>,
+    mut reader: Mp4Reader<File>,
+    sender: Sender<EncodedChunk>,
+    stop_thread: Arc<AtomicBool>,
+    length_size: u8,
+    TrackInfo {
+        sample_count,
+        track_id,
+        timescale,
+    }: TrackInfo,
+) {
+    let mut previous_sent_ts = None;
+    for i in 1..sample_count {
+        match reader.read_sample(track_id, i) {
+            Ok(Some(sample)) => {
+                let mut sample_data = sample.bytes.reader();
+                let mut data: BytesMut = Default::default();
+
+                if let Some(first_nal) = sps_and_pps.take() {
+                    data.extend_from_slice(&first_nal);
+                }
+
+                // the mp4 sample contains one h264 access unit (possibly more than one NAL).
+                // the NALs are stored as: <length_size bytes long big endian encoded length><the NAL>.
+                // we need to convert this into Annex B, in which NALs are separated by
+                // [0, 0, 0, 1]. `lenght_size` is at most 4 bytes long.
+                loop {
+                    let mut len = [0u8; 4];
+
+                    if sample_data
+                        .read_exact(&mut len[4 - length_size as usize..])
+                        .is_err()
+                    {
+                        break;
+                    }
+
+                    let len = u32::from_be_bytes(len);
+
+                    let mut nalu = bytes::BytesMut::zeroed(len as usize);
+                    sample_data.read_exact(&mut nalu).unwrap();
+
+                    data.extend_from_slice(&[0, 0, 0, 1]);
+                    data.extend_from_slice(&nalu);
+                }
+
+                let chunk = EncodedChunk {
+                    data: data.freeze(),
+                    pts: Duration::from_secs_f64(
+                        (sample.start_time as f64 + sample.rendering_offset as f64)
+                            / timescale as f64,
+                    ),
+                    dts: Some(Duration::from_secs_f64(
+                        sample.start_time as f64 / timescale as f64,
+                    )),
+                    kind: EncodedChunkKind::Video(VideoCodec::H264),
+                };
```

Comment:
```suggestion
                let dts = Duration::from_secs_f64(sample.start_time as f64 / timescale as f64)
                let chunk = EncodedChunk {
                    data: data.freeze(),
                    pts: Duration::from_secs_f64(
                        (sample.start_time as f64 + sample.rendering_offset as f64)
                            / timescale as f64,
                    ),
                    dts: Some(dts),
                    kind: EncodedChunkKind::Video(VideoCodec::H264),
                };
```
and use that dts value instead of unwrapping bellow